### DISCUSSION TITLE
Endpoints as first-class functionalities and Documentation on Services

### DISCUSSION THREAD

**Discussion by VaZark** [directus/directus#11632] [Feature Requests]
Currently, Directus treats Custom API as a part of the extension system. While this might be architecturally easier to maintain, applications that are not purely CMS routinely require custom routes for functionality. It would be nicer if creating custom routes are introduced in the Quickstart/App guide and not just part of the Extension system.

Additionally, it'd be interesting if the Settings area lists the registered routes and allowed admins to test the endpoints or redirect to a swagger page (if that can be implemented or is hosted separately). 

Moreover, I'm not able to find detailed docs on how to use the services available to extensions. I'll be glad to write up some documentation if someone can guide me through it

**Comment by benhaynes** [comment]
Thanks for this! I leave it up to the engineers to respond, but am wondering if we could use some upcoming features (Waterways) to build/assign data to register new endpoints. 🤔 

CC @erondpowell for the Docs

**Comment by VaZark** [comment]
I'm not sure what _Waterways_ is ^^. 

Honestly, this proposal is to make it easier for me to pitch directus as a complete backend for single-tenant services so that it can be used for both client-facing API and internal dashboards. Something along the lines of batteries-included Node framework and not just another advanced CMS. (Maybe I'm imagining/misinterpreting the goals than that of the core team)

I was imagining that directus could implement something like the schema below for its routes. In the App, under a **Routes** module, each route is listed as
```
 "/route"
 - Uses Collections : ["Collection_1", "Collection_2"]
 - Uses Collection fields : { "Collection_1":["*"], "Collection_2":["f3", "f2","f1_*"]}

## If possible
 - Supports GraphQL
 - Supports REST
   - GET : Expected Input & Output Schema
   - SEARCH : Expected Input & Output Schema
   - POST : Expected Input & Output Schema
```

The values for collections and fields can be autogenerated for the default routes, while the custom routes can declare them as part of the router register. 

**Advantages:**

1. If these can be implemented generically by the base routes, custom routes can reuse and extend this description which will allow them to be scanned and displayed on App too. 
2. It ensures the app knows the routes require certain collections & fields to function and can throw a 500 even before executing logic that is missing data.
3. It can later be extended in the app to apply permission to routes themselves based on user roles.

**Reply by rijkvanzanten** [reply]
I believe what you're asking for here is effectively what we're doing in the /server/specs endpoints. You can already download the API specifications in OpenAPI and GraphQL formats, with TypeScript hopefully being support soon 👍🏻

**Reply by VaZark** [reply]
That is partially what I meant. However, I want to define the spec to the custom endpoints as well. 

This question arose because, in the docs for [endpoints](https://docs.directus.io/extensions/endpoints/), I am not able to find any notes on how to define the spec for custom endpoints as well

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
